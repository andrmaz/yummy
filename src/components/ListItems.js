import React, { lazy, memo, Suspense } from 'react';
import PropTypes from 'prop-types';
import { SimpleGrid, Alert, AlertIcon, Spinner } from '@chakra-ui/react';

const CardItem = lazy(() => import('./CardItem'));

// Check on props change
export function areEqual(prevProps, nextProps){
    if (prevProps === nextProps) {
        return true;
    } else {
        return false;
    }
};

const ListItems = ({ results }) => (
    <Suspense  
        fallback={
        <Spinner
            label='Loading'
            thickness="4px"
            speed="0.65s"
            emptyColor="gray.200"
            color="blue.500"
            size="xl"
            pos='absolute'
            top='50%'
            left='50%'
        />
        }
    >
        <SimpleGrid minChildWidth="300px" spacing="5px">
            {
                !results ?
                <Alert status="info">
                    <AlertIcon />
                    Welcome to Yummy. Let&apos;s start searching!
                </Alert> :
                (
                    results?.count > 0 ?
                    results.hits.map(({recipe}) => (
                        <CardItem key={recipe.label} recipe={recipe} />
                    )) : 
                    <Alert status="warning">
                        <AlertIcon />
                        No items were found matching your search request .. Please retry
                    </Alert>
                )
            }
        </SimpleGrid>
    </Suspense>
);

ListItems.propTypes = {
    results: PropTypes.object
};

export default memo(ListItems, areEqual);
