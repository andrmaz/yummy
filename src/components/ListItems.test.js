import React from 'react';
import { shallow } from 'enzyme';
import ListItems, { areEqual } from './ListItems';

let wrapper;
const mockResults = {
    count: 2,
    hits: [{}, {}]
};

describe('ListItems Component', () => {
    beforeEach(() => {
        expect.hasAssertions();
    });
    it('should render Component correctly', () => {
        wrapper = shallow(<ListItems results={mockResults} />);
        expect(wrapper).toMatchSnapshot();
    });
    it('should receive an object correctly', () => {
        wrapper = shallow(<ListItems results={mockResults} />);
        expect(wrapper.prop('results')).toHaveProperty('count', 2);
    });
    it('should display welcome message at first', () => {
        wrapper = shallow(<ListItems results={undefined} />);
        expect(wrapper.text()).toMatch(/Welcome to Yummy/);
    });
    it('should display retry message if there are no matching results', () => {
        wrapper = shallow(<ListItems results={{ count: 0 }} />);
        expect(wrapper.text()).toMatch(/retry/);
    });
    it('should return true if props do not change', () => {
        expect(areEqual(1, 1)).toBeTruthy();
    });
    it('should return false on props change', () => {
        expect(areEqual(1, 0)).toBeFalsy();
    });
});
