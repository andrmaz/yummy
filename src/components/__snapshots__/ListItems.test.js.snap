// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItems Component should render Component correctly 1`] = `
<Suspense
  fallback={
    <Spinner
      color="blue.500"
      emptyColor="gray.200"
      label="Loading"
      left="50%"
      pos="absolute"
      size="xl"
      speed="0.65s"
      thickness="4px"
      top="50%"
    />
  }
  results={
    Object {
      "count": 2,
      "hits": Array [
        Object {},
        Object {},
      ],
    }
  }
>
  <SimpleGrid
    minChildWidth="300px"
    spacing="5px"
  >
    <lazy
      key="0"
    />
    <lazy
      key="1"
    />
  </SimpleGrid>
</Suspense>
`;
